File: cs/formula-cs/FormulaTest/DataContextTest.cs
using Formula;
using FormulaTest.Assertions;

namespace FormulaTest;

using static ResolvedValueAssertions;

public class DataContextTest
{
    [Test, TestCaseSource(typeof(DataContextTestCaseSource), nameof(DataContextTestCaseSource.GetTestCases))]
    public void DataDrivenTestCases(DataContextTestCaseSource.Case data)
    {
        var context = data.Data != null ? new DataContext(data.Data) : new DataContext();

        foreach (var dataActionText in data.Actions)
        {
            Execute(dataActionText, context);
        }
        
        foreach (var (key, value) in data.Expected)
        {
            if (value.ExpectedText != null)
            {
                AssertResolvedValue(context.Get(key)).HasValue(value.ExpectedText);
            }
        }
    }

    private void Execute(string actionText, DataContext context)
    {
        var parts = actionText.Split(" ");
        if (parts is ["SET", _, _])
        {
            context.Set(parts[1], ParseValue(parts[2]));
        }
        if (parts is ["PUSH", _, _])
        {
            context.Push(parts[1], ParseValue(parts[2]));
        }
    }

    private static IResolvable ParseValue(string text)
    {
        if (text.StartsWith("{") && text.EndsWith("}"))
        {
            return Formula.Formula.Parse(text.Substring(1, text.Length - 2)) ?? Resolvable.Empty;
        }

        return Resolvable.Just(text);
    }
}

File: js/formula-js/src/DataContext.test.ts
import {Formula} from "./Formula";
import * as fs from "node:fs";
import * as path from "node:path";
import {DataContext, MutableDataContext} from "./DataContext";
import {Resolvable} from "./Resolvable";
import {ResolvedValue} from "./ResolvedValue";

const YAML = require("yaml");

const dataContextTestCasesPath = path.resolve(__dirname, '..', '..', 'formula-test', 'data-context-test-cases.yml');
const dataContextTestCasesYaml = fs.readFileSync(dataContextTestCasesPath, "utf8");

const dataContextTestCases = YAML.parse(dataContextTestCasesYaml);

dataContextTestCases.forEach(testCase => {

  test(testCase.name, () => {
    const context = DataContext.of({});
    for (let actionText of testCase.actions) {
      const action = parseDataContextAction(actionText);
      action(context);
    }

    Object.keys(testCase.expected).forEach(key => {
      const expectedForKey = testCase.expected[key];
      const actualValue = context.get(key);
      expectValues(actualValue, expectedForKey);
    })
  });
});

function expectValues(actualValue: ResolvedValue, expectedForKey: any) {
  if ('expected_text' in expectedForKey) {
    expect(actualValue.asText()).toEqual(expectedForKey['expected_text']);
  }
  if ('expected_number' in expectedForKey) {
    expect(actualValue.asNumber()).toEqual(expectedForKey['expected_number']);
  }
  if ('expected_boolean' in expectedForKey) {
    expect(actualValue.asBoolean()).toEqual(expectedForKey['expected_boolean']);
  }
  if ('expected_list' in expectedForKey) {
    for (let i = 0; i < expectedForKey['expected_list'].length; i++) {
      expectValues(actualValue.asList()[i], expectedForKey['expected_list'][i]);
    }
  }
}

function toDataContext(data: any): DataContext {
  if (!data) return undefined;
  const parsed = {};
  Object.keys(data).forEach(key => {
    parsed[key] = parseDataValue(data[key]);
  });
  return DataContext.of(parsed);
}

function parseDataContextAction(text: string): (context: MutableDataContext) => any {
  return (context: MutableDataContext) => {
    const parts = text.split(' ');
    const action = parts[0].toUpperCase();
    const key = parts[1];
    const value = parts[2];

    switch (action) {
      case "SET":
        context.set(key, parseDataValue(value));
        break;
      case "PUSH":
        context.push(key, Resolvable.just(value));
        break;
    }
  };
}

function parseDataValue(value: any): string|number|boolean|ResolvedValue[]|Resolvable {
  if (typeof value === "string"
      && value.startsWith("{")
      && value.endsWith("}")) {
    return Formula.parse(value.slice(1, -1));
  }
  if (Array.isArray(value)) {
    return value.map(ResolvedValue.of);
  }
  return value;
}